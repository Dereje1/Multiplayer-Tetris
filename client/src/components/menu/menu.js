import React from 'react';
import PropTypes from 'prop-types';
import { NavLink, withRouter } from 'react-router-dom';
/* font awesome */
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faBars, faTimes, faSignOutAlt, faUser, faGamepad,
} from '@fortawesome/free-solid-svg-icons';
import './styles/menu.css';

class Menu extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      showMenu: false,
      delay: false,
    };
  }

  componentDidMount() {
    window.addEventListener('click', () => this.closeMenu());
  }

  componentDidUpdate(prevprops) {
    const { location: { pathname: currentPath } } = this.props;
    const { location: { pathname: prevPath } } = prevprops;
    if (prevPath !== currentPath) this.closeMenu();
  }

  componentWillUnmount() {
    window.removeEventListener('click');
    clearTimeout(this.closeId);
  }

  openMenu = (e) => {
    this.setState({ showMenu: true });
    e.stopPropagation();
  }

  closeMenu = () => {
    const { showMenu } = this.state;
    if (!showMenu) return;
    // added delay so slide out css effect can be seen
    this.setState({ delay: true }, () => {
      this.closeId = setTimeout(() => {
        clearTimeout(this.closeId);
        this.setState({ showMenu: false, delay: false });
      }, 475);
    });
  }

  render() {
    const { showMenu, delay } = this.state;
    const { onLogOut } = this.props;

    if (!showMenu && !delay) {
      return (
        <FontAwesomeIcon
          className="burger"
          icon={faBars}
          onClick={e => this.openMenu(e)}
        />
      );
    }

    return (
      <div
        id="menu"
        className={!delay ? 'show' : 'hide'}
        onKeyDown={() => {}}
        tabIndex={-1}
        role="menuitem"
        onClick={e => this.openMenu(e)}
      >
        <FontAwesomeIcon
          className="close"
          icon={faTimes}
          onClick={() => this.closeMenu()}
        />
        <div id="items">
          <div
            className="menuitems"
            onClick={() => onLogOut()}
            onKeyDown={() => {}}
            role="menuitem"
            tabIndex={-1}
          >
            <FontAwesomeIcon
              icon={faSignOutAlt}
            />
            <p>Logout</p>
          </div>
          <NavLink className="menuitems" exact to="/profile">
            <FontAwesomeIcon
              icon={faUser}
            />
            <p>Profile</p>
          </NavLink>
          <div>
            <NavLink className="menuitems" exact to="/">
              <FontAwesomeIcon
                icon={faGamepad}
              />
              <p>Game</p>
            </NavLink>
          </div>
        </div>
      </div>);
  }

}

export default withRouter(Menu);

Menu.propTypes = {
  onLogOut: PropTypes.func.isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
  }).isRequired,
};
