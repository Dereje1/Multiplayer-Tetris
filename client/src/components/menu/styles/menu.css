#menu {
  position: fixed;
  top: 0;
  left: 0;
  width: 225px;
  min-height: 100vh;
  background: rgba(67, 110, 89, 0.93);
}

#menu .close {
  font-size: 30px;
  text-align: right;
  position: absolute;
  right: 0;
  margin: 10px;
  cursor: pointer;
}

#menu.show {
  -webkit-animation-name: slideout;
          animation-name: slideout;
  -webkit-animation-duration: .5s;
          animation-duration: .5s;
  -webkit-animation-timing-function: linear;
          animation-timing-function: linear;
  -webkit-animation-iteration-count: 1;
          animation-iteration-count: 1;
}

#menu.hide {
  -webkit-animation-name: slidein;
          animation-name: slidein;
  -webkit-animation-duration: .5s;
          animation-duration: .5s;
  -webkit-animation-timing-function: linear;
          animation-timing-function: linear;
  -webkit-animation-iteration-count: 1;
          animation-iteration-count: 1;
}

#menu #items {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
  margin-top: 50%;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
}

#menu #items .menuitems {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
      -ms-flex-direction: row;
          flex-direction: row;
  -ms-flex-pack: distribute;
      justify-content: space-around;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  font-family: 'Raleway', sans-serif;
  color: #f0ecec;
  font-size: 1.5em;
  font-weight: bold;
  cursor: pointer;
  width: 200px;
  text-decoration: none;
}

#menu #items .menuitems.active {
  color: #2aff00;
  font-weight: bold;
}

#menu #items :hover {
  color: #2aff00;
}

.burger {
  position: absolute;
  top: 20px;
  left: 0;
  font-size: 40px;
  margin-left: 20px;
  cursor: pointer;
}

@-webkit-keyframes slideout {
  0% {
    -webkit-transform: translateX(-225px);
            transform: translateX(-225px);
  }
  50% {
    -webkit-transform: translateX(-100px);
            transform: translateX(-100px);
  }
  100% {
    -webkit-transform: translateX(0px);
            transform: translateX(0px);
  }
}

@keyframes slideout {
  0% {
    -webkit-transform: translateX(-225px);
            transform: translateX(-225px);
  }
  50% {
    -webkit-transform: translateX(-100px);
            transform: translateX(-100px);
  }
  100% {
    -webkit-transform: translateX(0px);
            transform: translateX(0px);
  }
}

@-webkit-keyframes slidein {
  0% {
    -webkit-transform: translateX(0px);
            transform: translateX(0px);
  }
  50% {
    -webkit-transform: translateX(-100px);
            transform: translateX(-100px);
  }
  100% {
    -webkit-transform: translateX(-225px);
            transform: translateX(-225px);
  }
}

@keyframes slidein {
  0% {
    -webkit-transform: translateX(0px);
            transform: translateX(0px);
  }
  50% {
    -webkit-transform: translateX(-100px);
            transform: translateX(-100px);
  }
  100% {
    -webkit-transform: translateX(-225px);
            transform: translateX(-225px);
  }
}

@media only screen and (max-width: 585px) {
  .burger, #menu {
    visibility: hidden;
  }
}
/*# sourceMappingURL=menu.css.map */