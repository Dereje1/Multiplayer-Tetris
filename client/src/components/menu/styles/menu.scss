#menu{
    position: fixed;
    top: 0;
    left: 0;
    width: 225px;
    min-height: 100vh;
    background: rgba(67, 110, 89, 0.93);
    z-index: 1000;
    .close{
        font-size: 30px;
        text-align: right;
        position: absolute;
        right: 0;
        margin: 10px;
        cursor: pointer;
    }
    &.show{
        animation-name: slideout;
        animation-duration: .5s;
        animation-timing-function: linear;
        animation-iteration-count: 1;
    }
    &.hide{
        animation-name: slidein;
        animation-duration: .5s;
        animation-timing-function: linear;
        animation-iteration-count: 1;
    }
    #items{
        display: flex;
        flex-direction: column;
        margin-top: 50%;
        align-items: center;
        .menuitems{
            display: flex;
            flex-direction: row;
            justify-content: space-around;
            align-items: center;
            font-family: 'Raleway', sans-serif;
            color: rgb(240, 236, 236);
            font-size: 1.5em;
            font-weight: bold;
            cursor: pointer;
            width: 200px;
            text-decoration: none;
            &.active{
                color: #2aff00;
                font-weight: bold;
            }
        }
        :hover{
            color:#2aff00;
        }
    }
}

.burger{
    position: absolute;
    top:20px;
    left:0;
    font-size: 40px;
    margin-left: 20px;
    cursor: pointer;
}

@keyframes slideout {
    0%{
      transform: translateX(-225px);
    }
    50%{
      transform: translateX(-100px);
    }
    100%{
      transform: translateX(0px);
    }
  }

@keyframes slidein {
    0%{
        transform: translateX(0px);
    }
    50%{
        transform: translateX(-100px);
    }
    100%{
        transform: translateX(-225px);
    }
}

@media only screen and (max-width: 585px) {
    .burger, #menu {
        visibility: hidden;
    }
}