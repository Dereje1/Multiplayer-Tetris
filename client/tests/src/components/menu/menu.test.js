/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Menu } from '../../../../src/components/menu/menu'

jest.useFakeTimers();
describe('The Menu', () => {
    let props, closeMenuSpy;
    beforeEach(() => {
        props = {
            onLogOut: jest.fn(),
            location: { pathname: '/' }
        }
        jest.spyOn(window, 'addEventListener').mockImplementationOnce(() => { });
        jest.spyOn(window, 'removeEventListener').mockImplementationOnce(() => { });
    })
    afterEach(() => {
        props = null
        window.addEventListener.mockClear()
        window.removeEventListener.mockClear()
        closeMenuSpy.mockClear()
    })
    test('will render the collapsed menu', () => {
        const wrapper = shallow(<Menu {...props} />)
        closeMenuSpy = jest.spyOn(wrapper.instance(), 'closeMenu');
        const [listener, callback] = window.addEventListener.mock.calls[0]
        callback()
        expect(listener).toBe('click')
        expect(closeMenuSpy).toHaveBeenCalled()
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    test('will render the expanded menu', () => {
        const wrapper = shallow(<Menu {...props} />)
        expect(wrapper.state().showMenu).toBe(false)
        wrapper.props().onClick({ stopPropagation: jest.fn() })
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.state().showMenu).toBe(true)
    })

    test('will close the expanded menu', () => {
        const wrapper = shallow(<Menu {...props} />)
        wrapper.props().onClick({ stopPropagation: jest.fn() })
        expect(wrapper.state().showMenu).toBe(true)
        const closeIcon = wrapper.find({ className: "close"})
        closeIcon.props().onClick()
        jest.advanceTimersByTime(475);
        expect(wrapper.state().showMenu).toBe(false)
    })

    test('will close the menu on a path change',()=>{
        const wrapper = shallow(<Menu {...props} />)
        closeMenuSpy = jest.spyOn(wrapper.instance(), 'closeMenu');
        wrapper.instance().componentDidUpdate({ location: { pathname: '/otherpath' } })
        expect(closeMenuSpy).toHaveBeenCalled()
    })

    test('will logout the user', () => {
        const wrapper = shallow(<Menu {...props} />)
        wrapper.props().onClick({ stopPropagation: jest.fn() })
        
        const logoutIcon = wrapper.find({ className: "menuitems"}).at(0)
        logoutIcon.props().onClick();
        expect(props.onLogOut).toHaveBeenCalled()
    })

    test('will remove the listener whem menu unmounts',()=>{
        const wrapper = shallow(<Menu {...props} />)
        wrapper.instance().componentWillUnmount()
        expect(window.removeEventListener).toHaveBeenCalledWith('click')
    })
})