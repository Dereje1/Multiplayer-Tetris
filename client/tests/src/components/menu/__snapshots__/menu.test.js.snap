// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Menu will render the collapsed menu 1`] = `
<FontAwesomeIcon
  beat={false}
  beatFade={false}
  border={false}
  bounce={false}
  className="burger"
  fade={false}
  fixedWidth={false}
  flip={false}
  icon={
    Object {
      "icon": Array [
        448,
        512,
        Array [],
        "f0c9",
        "M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z",
      ],
      "iconName": "bars",
      "prefix": "fas",
    }
  }
  inverse={false}
  listItem={false}
  mask={null}
  maskId={null}
  onClick={[Function]}
  pull={null}
  pulse={false}
  rotation={null}
  shake={false}
  size={null}
  spin={false}
  spinPulse={false}
  spinReverse={false}
  swapOpacity={false}
  symbol={false}
  title=""
  titleId={null}
  transform={null}
/>
`;

exports[`The Menu will render the expanded menu 1`] = `
<div
  className="show"
  id="menu"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="menuitem"
  tabIndex={-1}
>
  <FontAwesomeIcon
    beat={false}
    beatFade={false}
    border={false}
    bounce={false}
    className="close"
    fade={false}
    fixedWidth={false}
    flip={false}
    icon={
      Object {
        "icon": Array [
          352,
          512,
          Array [],
          "f00d",
          "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
        ],
        "iconName": "times",
        "prefix": "fas",
      }
    }
    inverse={false}
    listItem={false}
    mask={null}
    maskId={null}
    onClick={[Function]}
    pull={null}
    pulse={false}
    rotation={null}
    shake={false}
    size={null}
    spin={false}
    spinPulse={false}
    spinReverse={false}
    swapOpacity={false}
    symbol={false}
    title=""
    titleId={null}
    transform={null}
  />
  <div
    id="items"
  >
    <div
      className="menuitems"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="menuitem"
      tabIndex={-1}
    >
      <FontAwesomeIcon
        beat={false}
        beatFade={false}
        border={false}
        bounce={false}
        className=""
        fade={false}
        fixedWidth={false}
        flip={false}
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f2f5",
              "M497 273L329 441c-15 15-41 4.5-41-17v-96H152c-13.3 0-24-10.7-24-24v-96c0-13.3 10.7-24 24-24h136V88c0-21.4 25.9-32 41-17l168 168c9.3 9.4 9.3 24.6 0 34zM192 436v-40c0-6.6-5.4-12-12-12H96c-17.7 0-32-14.3-32-32V160c0-17.7 14.3-32 32-32h84c6.6 0 12-5.4 12-12V76c0-6.6-5.4-12-12-12H96c-53 0-96 43-96 96v192c0 53 43 96 96 96h84c6.6 0 12-5.4 12-12z",
            ],
            "iconName": "sign-out-alt",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        maskId={null}
        pull={null}
        pulse={false}
        rotation={null}
        shake={false}
        size={null}
        spin={false}
        spinPulse={false}
        spinReverse={false}
        swapOpacity={false}
        symbol={false}
        title=""
        titleId={null}
        transform={null}
      />
      <p>
        Logout
      </p>
    </div>
    <NavLink
      className="menuitems"
      exact={true}
      onClick={[Function]}
      to="/profile"
    >
      <FontAwesomeIcon
        beat={false}
        beatFade={false}
        border={false}
        bounce={false}
        className=""
        fade={false}
        fixedWidth={false}
        flip={false}
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f007",
              "M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z",
            ],
            "iconName": "user",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        maskId={null}
        pull={null}
        pulse={false}
        rotation={null}
        shake={false}
        size={null}
        spin={false}
        spinPulse={false}
        spinReverse={false}
        swapOpacity={false}
        symbol={false}
        title=""
        titleId={null}
        transform={null}
      />
      <p>
        Profile
      </p>
    </NavLink>
    <div>
      <NavLink
        className="menuitems"
        exact={true}
        onClick={[Function]}
        to="/"
      >
        <FontAwesomeIcon
          beat={false}
          beatFade={false}
          border={false}
          bounce={false}
          className=""
          fade={false}
          fixedWidth={false}
          flip={false}
          icon={
            Object {
              "icon": Array [
                640,
                512,
                Array [],
                "f11b",
                "M480.07 96H160a160 160 0 1 0 114.24 272h91.52A160 160 0 1 0 480.07 96zM248 268a12 12 0 0 1-12 12h-52v52a12 12 0 0 1-12 12h-24a12 12 0 0 1-12-12v-52H84a12 12 0 0 1-12-12v-24a12 12 0 0 1 12-12h52v-52a12 12 0 0 1 12-12h24a12 12 0 0 1 12 12v52h52a12 12 0 0 1 12 12zm216 76a40 40 0 1 1 40-40 40 40 0 0 1-40 40zm64-96a40 40 0 1 1 40-40 40 40 0 0 1-40 40z",
              ],
              "iconName": "gamepad",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          maskId={null}
          pull={null}
          pulse={false}
          rotation={null}
          shake={false}
          size={null}
          spin={false}
          spinPulse={false}
          spinReverse={false}
          swapOpacity={false}
          symbol={false}
          title=""
          titleId={null}
          transform={null}
        />
        <p>
          Game
        </p>
      </NavLink>
    </div>
  </div>
</div>
`;
